### PR TITLE
ARROW-11318: [Rust] Support pretty printing timestamp, date, and timestamp types

### PR DESCRIPTION
**Pull request by alamb**
I found this while removing `test::format_batches` (PR to come shortly); The Record batch pretty printing code was printing numbers rather than dates.

Before this PR, when date/time columns were printed they were printed as numbers:

```
[
    "+----------+",
    "| f        |",
    "+----------+",
    "| 11111111 |",
    "|          |",
    "+----------+",
]
```

After this PR, they are printed (via chrono) as dates:

```
[
    "+---------------------+",
    "| f                   |",
    "+---------------------+",
    "| 1970-05-09 14:25:11 |",
    "|                     |",
    "+---------------------+",
]
```

### CODE REVIEW

File: rust/arrow/src/util/display.rs [hunk 1]
```diff
@@ -104,30 +155,30 @@ pub fn array_value_to_string(column: &array::ArrayRef, row: usize) -> Result<Str
...
-            make_string!(array::Time32SecondArray, column, row)
+            make_string_time!(array::Time32SecondArray, column, row)
         }
         DataType::Time32(unit) if *unit == TimeUnit::Millisecond => {
-            make_string!(array::Time32MillisecondArray, column, row)
+            make_string_time!(array::Time32MillisecondArray, column, row)
         }
-        DataType::Time32(unit) if *unit == TimeUnit::Microsecond => {
-            make_string!(array::Time64MicrosecondArray, column, row)
+        DataType::Time64(unit) if *unit == TimeUnit::Microsecond => {
```
Comment (alamb) [inline rust/arrow/src/util/display.rs:177]: There was actually a bug here -- there is no such type as using `Time32` for storing Microseconds -- it is actually `Time64` (as can be seen in the `Time64MicrosecondArray` type below it)

File: rust/arrow/src/util/display.rs [hunk 2]
```diff
@@ -44,6 +44,57 @@ macro_rules! make_string {
...
+    ($array_type:ty, $column: ident, $row: ident) => {{
+        let array = $column.as_any().downcast_ref::<$array_type>().unwrap();
+
+        let s = if array.is_null($row) {
+            "".to_string()
+        } else {
+            array
+                .value_as_date($row)
+                .map(|d| d.to_string())
+                .unwrap_or_else(|| String::from("ERROR CONVERTING DATE"))
```
Comment (Dandandan) [inline rust/arrow/src/util/display.rs:57]:
```suggestion
                .unwrap_or_else(|| "ERROR CONVERTING DATE".to_string())
```

File: rust/arrow/src/util/display.rs [hunk 3]
```diff
@@ -44,6 +44,57 @@ macro_rules! make_string {
...
+    ($array_type:ty, $column: ident, $row: ident) => {{
+        let array = $column.as_any().downcast_ref::<$array_type>().unwrap();
+
+        let s = if array.is_null($row) {
+            "".to_string()
+        } else {
+            array
+                .value_as_time($row)
+                .map(|d| d.to_string())
+                .unwrap_or_else(|| String::from("ERROR CONVERTING TIME"))
```
Comment (Dandandan) [inline rust/arrow/src/util/display.rs:74]:
```suggestion
                .unwrap_or_else(|| "ERROR CONVERTING DATE".to_string())
```

File: rust/arrow/src/util/display.rs [hunk 4]
```diff
@@ -44,6 +44,57 @@ macro_rules! make_string {
...
+    ($array_type:ty, $column: ident, $row: ident) => {{
+        let array = $column.as_any().downcast_ref::<$array_type>().unwrap();
+
+        let s = if array.is_null($row) {
+            "".to_string()
+        } else {
+            array
+                .value_as_datetime($row)
+                .map(|d| d.to_string())
+                .unwrap_or_else(|| String::from("ERROR CONVERTING DATETIME"))
```
Comment (Dandandan) [inline rust/arrow/src/util/display.rs:91]:
```suggestion
                .unwrap_or_else(|| "ERROR CONVERTING DATE".to_string())
```

**Review by paddyhoran — APPROVED** [review]

File: rust/arrow/src/util/display.rs [hunk 5]
```diff
@@ -44,6 +44,57 @@ macro_rules! make_string {
     }};
 }
 
+macro_rules! make_string_date {
+    ($array_type:ty, $column: ident, $row: ident) => {{
+        let array = $column.as_any().downcast_ref::<$array_type>().unwrap();
+
+        let s = if array.is_null($row) {
+            "".to_string()
```
[the remark this answers is not in the document]
Comment (alamb) [reply rust/arrow/src/util/display.rs:52]: I think there are tradeoffs between using empty string or some sentinel (`"NULL"`) for display; I don't have a super strong preference though I do think the behavior  should be consistent across types

 This PR follows the same convention as the rest of this module and uses "" for NULL values: https://github.com/apache/arrow/pull/9263/files#diff-821be3a8a9239cdd12ab2b3c979e174f7b936aa6bfad4b9bc76489b9923cf747R38

[the remark this answers is not in the document]
Comment (alamb) [reply rust/arrow/src/util/display.rs:52]: The NULL behavior is included in the tests below (and the empty string is visible)

[the remark this answers is not in the document]
Comment (Dandandan) [reply rust/arrow/src/util/display.rs:52]: Makes sense to me. I think the "problem"  would be when displaying strings, as the empty string is displayed the same way as null, but I agree it is better to keep it consistent.

**Review by Dandandan — APPROVED** [review]

**Pull request closed without merging**